src/ui/Icon/Icon.tsx

```diff
@@ -0,0 +1,76 @@
+import camelCase from 'lodash/camelCase'
+
+import * as svgDeveloper from './svg/developer'
+import * as svgOutline from './svg/outline'
+import * as svgSolid from './svg/solid'
+
+const iconComponentCollection = {
+  outline: svgOutline,
+  solid: svgSolid,
+  developer: svgDeveloper,
+}
+
+type IconCollection = typeof iconComponentCollection
+type Variant = keyof IconCollection
+
+type OutlineIconCollection = typeof svgOutline
+type SolidIconCollection = typeof svgSolid
+type DeveloperIconCollection = typeof svgDeveloper
+
+type Name = keyof (OutlineIconCollection &
+  SolidIconCollection &
+  DeveloperIconCollection)
+
+function isValidKey<T extends Record<string, any>>(
+  obj: T,
+  key: keyof any
+): key is keyof T {
+  return key in obj
+}
+
+function get(
+  iconCollection: IconCollection,
+  type: Variant,
+  name: Name
+): React.FunctionComponent<React.SVGAttributes<SVGElement>> | undefined {
+  const collection = iconCollection[type]
+  const camelName = camelCase(name)
+
+  if (camelName && isValidKey(collection, camelName)) {
+    return collection[camelName]
+  }
+}
+
+const iconClasses = {
+  sm: 'w-4 h-4',
+  flex: 'w-5 h-5', // This doesn't look right?
+  md: 'w-6 h-6',
+  lg: 'w-16 h-16',
+}
+
+interface IconProps {
+  name: Name
+  variant?: Variant
+  size?: 'sm' | 'md' | 'lg' | 'flex'
+  label?: string
+}
+
+function Icon({
+  name,
+  variant = 'outline',
+  size = 'md',
+  label = '',
+}: IconProps) {
+  const IconSvg = get(iconComponentCollection, variant, name)
+  if (!IconSvg || !isValidKey(iconClasses, size)) return null
+
+  return (
+    <IconSvg
+      data-testid={label}
+      data-icon={label}
+      className={iconClasses[size]}
+    />
+  )
+}
```

Comment:
Just to take this to the next TypeScript level and make sure we don't introduce any errors because the variants don't have all of the same values for name, can you switch the props to the following so when you set the variant to `developer` you'll only see the names that are possible for it:

```tsx
interface CommonProps {
  size?: 'sm' | 'md' | 'lg' | 'flex'
  label?: string
}

type DeveloperIconProps = {
  name: keyof DeveloperIconCollection
  variant?: 'developer'
}

type OutlineIconProps = {
  name: keyof OutlineIconCollection
  variant?: 'outline'
}

type SolidIconProps = {
  name: keyof SolidIconCollection
  variant?: 'solid'
}

type IconProps = (DeveloperIconProps | OutlineIconProps | SolidIconProps) &
  CommonProps

function Icon({
  name,
  variant = 'outline',
  size = 'md',
  label = '',
}: IconProps) {
  const IconSvg = get(iconComponentCollection, variant, name)
  if (!IconSvg || !isValidKey(iconClasses, size)) return null

  return (
    <IconSvg
      data-testid={label}
      data-icon={label}
      className={iconClasses[size]}
    />
  )
}
```

Reply:
I wasn't able to get the type narrowing working correctly with `type IconProps = (DeveloperIconProps | OutlineIconProps | SolidIconProps) &
  CommonProps` but moving the intersection CommonProps up to each IconProps and unioning them seemed to have done the trick.
  
  ```
type OutlineIconProps = {
  name: keyof OutlineIconCollection
  variant: 'outline'
} & CommonProps
```

If anyone can why explain why this makes the difference to me that would be great. I would have assumed both ways were equivalent.